diffprivlib/accountant.py

```diff
@@ -0,0 +1,294 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2020
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Privacy budget accountant for differential privacy
+"""
+from numbers import Integral
+
+import numpy as np
+
+from diffprivlib.utils import check_epsilon_delta, BudgetError
+
+
+class BudgetAccountant:
+    """Privacy budget accountant for differential privacy.
+
+    This class creates a privacy budget accountant to track privacy spend across queries and other data accesses. Once
+    initialised, the BudgetAccountant stores each privacy spend and iteratively updates the total budget spend, raising
+    an error when the total budget (if specified) is exceeded. The accountant can be initialised without any maximum
+    budget, to enable users track the total privacy spend of their actions without hindrance.
+
+    Implements the accountant rules as given in [KOV17]_.
+
+    Parameters
+    ----------
+    epsilon : float, default: infinity
+        Target epsilon budget for the accountant.
+
+    delta : float, default : 1
+        Target delta budget for the accountant.
+
+    slack : float, default : 0
```

Comment:
I'm not sure if `slack` is the best thing to call this. In essence, it defines the slack that we're allowing in the overall delta budget. Ideas welcome!